inc/impact.class.php

```diff
@@ -0,0 +1,1091 @@
+<?php
+
+if (!defined('GLPI_ROOT')) {
+   die("Sorry. You can't access this file directly");
+}
+
+/**
+ * @since 9.5.0
+ */
+class Impact extends CommonGLPI {
+   // Constants used to express the direction or "flow" of a graph
+   // Theses constants can also be used to express if and edge is reachable
+   // when exploring the graph forward, backward or both (0b11)
+   const DIRECTION_FORWARD    = 0b01;
+   const DIRECTION_BACKWARD   = 0b10;
+
+   // Default colors used for the edges of the graph according to their flow
+   const DEFAULT_COLOR            = 'black';  // No flow, this mean the edge is not accessible from the starting point of the graph
+   const IMPACT_COLOR             = '#ff3418'; // Forward
+   const DEPENDS_COLOR            = '#1c76ff'; // Backward
+   const IMPACT_AND_DEPENDS_COLOR = '#ca29ff'; // Forward and backward
+
+   public static function getTypeName($nb = 0) {
+      return _n('Asset impact', 'Asset impacts', $nb);
+   }
+
+   public function getTabNameForItem(CommonGLPI $item, $withtemplate = 0) {
+      global $CFG_GLPI, $DB;
+
+      // Class of the current item
+      $class = get_class($item);
+
+      // Only enabled for CommonDBTM
+      if (!is_a($item, "CommonDBTM")) {
+         throw new InvalidArgumentException(
+            "Argument \$item ($class) must be a CommonDBTM)."
+         );
+      }
+
+      // If on an asset, get the number its direct dependencies
+      if (isset($CFG_GLPI['impact_asset_types'][$class])) {
+         $total = count($DB->request([
+            'FROM'   => ImpactRelation::getTable(),
+            'WHERE'  => [
+               'OR' => [
+                  [
+                     'itemtype_source' => get_class($item),
+                     'items_id_source' => $item->getID(),
+                  ],
+                  [
+                     'itemtype_impacted' => get_class($item),
+                     'items_id_impacted' => $item->getID(),
+                  ]
+               ]
+            ]
+         ]));
+      } else if (is_a($item, "CommonITILObject")) {
+         // Tab name for an ITIL object : always 0
+         $total = 0;
+      } else {
+         // Neither an asset nor an CommonITILObject
+         throw new InvalidArgumentException(
+            "Argument \$item ($class) is not a valid target for impact analysis."
+         );
+      }
+
+      return self::createTabEntry(
+         _n('Impact ', 'Impacts ', Session::getPluralNumber()),
+         $total
+      );
+   }
+
+   public static function displayTabContentForItem(
+      CommonGLPI $item,
+      $tabnum = 1,
+      $withtemplate = 0
+   ) {
+      global $CFG_GLPI;
+
+      $class = get_class($item);
+
+      // Only enabled for CommonDBTM
+      if (!is_a($item, "CommonDBTM")) {
+         throw new InvalidArgumentException(
+            "Argument \$item ($class) must be a CommonDBTM)."
+         );
+      }
+
+      $ID = $item->getID();
```

Comment:
Be very careful with `getID` usage (even if it looks like a good idea to use it, in facts this is not on many cases, but required in some others :/)

Reply:
Why ? What else can it be replaced with ?

Reply:
`$object->fields['id']`

Reply:
Fixed in latest commit.